src/libmongoc/src/mongoc/mongoc-topology.c

```diff
@@ -604,6 +638,8 @@ mongoc_topology_new (const mongoc_uri_t *uri, bool single_threaded)
       mongoc_topology_scanner_add (topology->scanner, elem, id, false);
    }
 
+   _detect_nongenuine_hosts (*hl_array);
```

Comment:
I realize this is an additional iteration (after `srvMaxHosts` is applied when applicable), but the resulting code is much cleaner than the previous commit where I incorporated the logic to log at-most-once during the preceding loop.

That seems preferable given that `mongoc_topology_new` is already quite complex and the only overhead here is following a handful of `mongoc_host_list_t` pointers.